<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Web facilities</title>

<script src="web_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="web_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="web_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="web_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="web_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="web_files/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="web_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="web_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="web_files/navigation-1.1/tabsets.js"></script>
<link href="web_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="web_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>





<style type="text/css">
/* for pandoc --citeproc since 2.11 */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">Web facilities</h1>

</div>


<div id="hello-world" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Hello, world!</h1>
<div id="python" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Python</h2>
<p>Perhaps this is the simplest way to start from the Linux command line, nevertheless with Python,</p>
<pre class="python"><code>python3 -m http.server 8000
firefox http://127.0.0.1:8000</code></pre>
<p>which works with <code>python3 -m http.server</code> by default when the port number 8000 is available.</p>
</div>
<div id="r" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> R</h2>
<p>Our focus here is R with the following script.</p>
<pre class="r"><code>httpuv::startServer(&quot;0.0.0.0&quot;, 8000, list(
  call = function(req) {
    list(
      status = 200L,
      headers = list(&quot;Content-Type&quot; = &quot;text/plain&quot;),
      body = &quot;Hello, world!&quot;
    )
  })
)</code></pre>
<p>Upon accessing <code>http://127.0.0.1:8000</code>, we see the message “Hello, world!”.</p>
</div>
</div>
<div id="url-availability" class="section level1" number="2">
<h1><span class="header-section-number">2</span> URL availability</h1>
<p>This is an attempt to implement the following CRAN policy,</p>
<blockquote>
<p>Packages which use Internet resources should fail gracefully with an informative message if the resource is not available or has changed (and not give a check warning nor error).</p>
</blockquote>
<pre class="r"><code>check_url_availability &lt;- function(url) {
# Send a GET request to the URL
  response &lt;- tryCatch({
    httr::GET(url)
  }, error = function(e) {
# If there was an error, return FALSE
    return(NULL)
  })
# If response is NULL or status code is not 200, return FALSE
  if (is.null(response) || httr::status_code(response) != 200) {
    message(paste(&quot;Warning: The URL&quot;, url, &quot;is not accessible. Please check the link.&quot;))
    return(FALSE)
  }
# If status code is 200, the URL is available
  message(paste(&quot;The URL&quot;, url, &quot;is accessible.&quot;))
  return(TRUE)
}

url_to_check &lt;- &quot;http://www.zotero.org/styles/nature-genetics&quot;
is_available &lt;- check_url_availability(url_to_check)
#&gt; The URL http://www.zotero.org/styles/nature-genetics is accessible.

if (is_available) {
  message(&quot;Using CSL as usual.&quot;)
} else {
  message(&quot;Using fallback to local CSL file instead.&quot;)
}
#&gt; Using CSL as usual.</code></pre>
<p>Below, we assume that our working directory is the source package,</p>
</div>
<div id="data-source" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Data source</h1>
<p>We envisage a server from which data can be obtained, and two R packages <strong>plumber</strong> and <strong>httpuv</strong> are considered here.</p>
<div id="a-meta-analysis" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> A meta-analysis</h2>
<p>The compressed data <code>IL.18R1-1.tbl.gz</code> is based on METAL<span class="citation"><sup>1</sup></span>, involving study-[1|2|3] as follows,</p>
<pre><code>SEPARATOR TAB
COLUMNCOUNTING STRICT
CHROMOSOMELABEL CHR
POSITIONLABEL POS
CUSTOMVARIABLE N
LABEL N as N
TRACKPOSITIONS ON
AVERAGEFREQ ON
MINMAXFREQ ON
ADDFILTER AF1 &gt;= 0.001
ADDFILTER AF1 &lt;= 0.999
MARKERLABEL SNP
ALLELELABELS A1 A2
EFFECTLABEL BETA
PVALUELABEL P
WEIGHTLABEL N
FREQLABEL AF1
STDERRLABEL SE
SCHEME STDERR
EFFECT_PRINT_PRECISION 8
STDERR_PRINT_PRECISION 8
GENOMICCONTROL OFF
LOGPVALUE ON
OUTFILE IL18R1B_dr- .tbl
PROCESS study-1-IL18R1B.fastGWA.gz
PROCESS study-2-IL18R1B.fastGWA.gz
PROCESS study-3-IL18R1B.fastGWA.gz
PROCESS study-1-IL18R1B-chrX.fastGWA.gz
PROCESS study-2-IL18R1B-chrX.fastGWA.gz
PROCESS study-3-IL18R1B-chrX.fastGWA.gz
ANALYZE HETEROGENEITY
CLEAR</code></pre>
<p>Assuming the script is named <code>IL.18R1.metal</code>, <code>IL.18R1-1.tbl.gz</code> is generated from <strong>htslib</strong>, <a href="https://www.htslib.org/download/" class="uri">https://www.htslib.org/download/</a>, as follows,</p>
<pre class="bash"><code>export protein=IL.18R1
metal ${protein}.metal 2&gt;&amp;1 | \
tee ${protein}-1.tbl.log
cat &lt;(head -1 ${protein}-1.tbl) \
    &lt;(sed &#39;1d&#39; ${protein}-1.tbl | \
sort -k1,1n -k2,2n) | \
bgzip -f &gt; ${protein}-1.tbl.gz
tabix -S1 -s1 -b2 -e2 -f ${protein}-1.tbl.gz
rm ${protein}-1.tbl</code></pre>
<p>The bgzipped data allows for access by genomic region as shown below.</p>
</div>
<div id="plumber" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> plumber</h2>
<p>Software library <code>libsodium</code>, <a href="https://doc.libsodium.org/" class="uri">https://doc.libsodium.org/</a>, is required for its installation.</p>
<p>It is <code>an API generator in R</code>, which has been tested as follows.</p>
<pre class="r"><code>get_data &lt;- function(filename, region)
{
  query_result &lt;- seqminer::tabix.read(filename, region)
  hdr &lt;- c(&quot;Chromosome&quot;, &quot;Position&quot;,
           &quot;MarkerName&quot;, &quot;Allele1&quot;, &quot;Allele2&quot;, &quot;Freq1&quot;, &quot;FreqSE&quot;, &quot;MinFreq&quot;, &quot;MaxFreq&quot;,
           &quot;Effect&quot;, &quot;StdErr&quot;, &quot;logP&quot;,
           &quot;Direction&quot;, &quot;HetISq&quot;, &quot;HetChiSq&quot;, &quot;HetDf&quot;, &quot;logHetP&quot;, &quot;N&quot;)
  df &lt;- read.table(text = paste(query_result, collapse = &quot;\n&quot;), sep = &quot;\t&quot;, col.names=hdr)
  return(df)
}

plbr &lt;- plumber::Plumber$new()
plbr$handle(&quot;GET&quot;, &quot;/tests&quot;, function(req, res) {
  protein &lt;- req$args$protein
  region &lt;- req$args$region
  if (is.null(protein) || is.null(region)) {
    res$status &lt;- 400
    return(list(error = &quot;Both &#39;protein&#39; and &#39;region&#39; must be provided&quot;))
  }
  filename &lt;- file.path(&quot;tests&quot;,paste0(protein,&quot;-1.tbl.gz&quot;))
  print(filename)
  if (!file.exists(filename)) {
    res$status &lt;- 404
    return(list(error = paste(&quot;File for&quot;, protein, &quot;not found&quot;)))
  }
  data &lt;- get_data(filename, region)
  json_data &lt;- jsonlite::toJSON(data, dataframe = &quot;rows&quot;, na = &quot;null&quot;)
  res$setHeader(&quot;Content-Type&quot;, &quot;application/json&quot;)
  return(json_data)
})
options(width = 200)
filename &lt;- file.path(&quot;tests&quot;,&quot;IL.18R1-1.tbl.gz&quot;)
region &lt;- &quot;2:102700000-103800000&quot;
data &lt;- get_data(filename, region)
head(data,1)
plbr$run(port = 8001)</code></pre>
<p>Indeed we can see that the first line of data,</p>
<pre><code>  Chromosome  Position         MarkerName Allele1 Allele2 Freq1 FreqSE MinFreq MaxFreq  Effect StdErr  logP   Direction HetISq HetChiSq HetDf logHetP     N
1          2 102700138 chr2:102700138_A_G       a       g 0.087 0.0207  0.0641  0.1376 -0.0566 0.0239 -1.75 -?-+n-?--n+   78.2   36.757     8  -4.894 12799</code></pre>
<p>and</p>
<pre><code>Running plumber API at http://127.0.0.1:8001
Running swagger Docs at http://127.0.0.1:8001/__docs__/</code></pre>
</div>
<div id="data-access" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Data access</h2>
<div id="browserconsole" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Browser/console</h3>
<p>So we get query results in JSON format from</p>
<ul>
<li><strong>browser</strong>: <code>http://localhost:8001/tests?protein=IL.18R1&amp;region=2:102700000-103800000</code></li>
<li><strong>command line interface</strong>: <code>curl "http://localhost:8001/tests?protein=IL.18R1&amp;region=2:102700000-103800000"</code></li>
</ul>
<p>Additional work required to get output from <code>curl</code> to a tab-delimited data,</p>
<pre class="bash"><code>curl &quot;http://localhost:8001/tests?protein=IL.18R1&amp;region=2:102700000-103800000&quot; | \
jq -r &#39;.[0] |
   fromjson |
   .[] |
   [
     .Chromosome, .Position, .MarkerName, .Allele1, .Allele2, .Freq1,
     .Effect, .StdErr, .logP, .Direction, .HetISq, .HetChiSq, .HetDf, .logHetP, .N
   ] |
   @tsv&#39;</code></pre>
<p>where</p>
<ol style="list-style-type: decimal">
<li><strong>.[0]</strong>: Access the first element in the outer array (the string containing the JSON).</li>
<li><strong>fromjson</strong>: Parse the string into a JSON object.</li>
<li><strong>.[]</strong>: Iterate over the array inside the parsed JSON.</li>
<li><strong>[ … ]</strong>: Create an array of the values needed, each corresponds to a column in the TSV output.</li>
<li><strong><span class="citation"><sup><strong>tsv?</strong></sup></span></strong>: Convert the array into tab-separated values.</li>
</ol>
<p>Note also that only selected columns (as in 4) are kept. The simplest way to have the header is add it manually,</p>
<pre class="bash"><code>(
  echo &quot;Chromosome|Position|MarkerName|Allele1|Allele2|Freq1|Effect|StdErr|logP|Direction|HetISq|HetChiSq|HetDf|logHetP|N&quot; | \
  sed &#39;s/|/\t/g&#39;
  curl command as above.
)</code></pre>
</div>
<div id="r-1" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> R</h3>
<p>The query above is easily furnished with <strong>curl</strong>:</p>
<pre class="r"><code>tmp &lt;- tempfile()
curl::curl_download(&quot;http://localhost:8001/tests?protein=IL.18R1&amp;region=2:102700000-103800000&quot;, tmp)
df &lt;- jsonlite::fromJSON(readLines(tmp)) |&gt;
      jsonlite::fromJSON(flatten=TRUE) |&gt;
      as.data.frame()
dim(df)</code></pre>
<p>giving</p>
<pre><code>[1] 4779   18</code></pre>
</div>
</div>
<div id="httpuv" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> httpuv</h2>
<p>The package gives a somewhat more involved version as follows,</p>
<pre class="r"><code>dir.create(&quot;content/assets&quot;, recursive = TRUE)
dir.create(&quot;content/lib&quot;, recursive = TRUE)
s &lt;- httpuv::startServer(
  host = &quot;0.0.0.0&quot;, 
  port = 5000,
  app = list(
    call = function(req) {
      list(
        status = 200L,
        headers = list(
          &#39;Content-Type&#39; = &#39;text/html&#39;,
          &#39;Access-Control-Allow-Origin&#39; = &#39;*&#39;,
          &#39;Access-Control-Allow-Methods&#39; = &#39;GET, POST, OPTIONS&#39;,
          &#39;Access-Control-Allow-Headers&#39; = &#39;Content-Type&#39;
        ),
        body = &quot;Hello world!&quot;
      )
    },
    staticPaths = list(
      &quot;/assets&quot; = &quot;content/assets/&quot;, # Static assets
      &quot;/lib&quot; = httpuv::staticPath(&quot;content/lib&quot;, indexhtml = FALSE),
      &quot;/lib/dynamic&quot; = httpuv::excludeStaticPath()
    ),
    staticPathOptions = httpuv::staticPathOptions(indexhtml = TRUE)
  )
)
cat(&quot;Server running at http://0.0.0.0:5000\n&quot;)
s$stop()</code></pre>
<p>so mappings are created from <code>content/[assets, lib]</code> to <code>assets</code> and <code>lib</code>, while <code>httpuv::excludeStaticPath()</code> indicates that requests
to <code>/lib/dynamic</code> will not be served as static files but could be handled dynamically by the app logic.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references csl-bib-body" line-spacing="2">
<div id="ref-willer10" class="csl-entry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Willer, C. J., Li, Y. &amp; Abecasis, G. R. <a href="https://doi.org/10.1093/bioinformatics/btq340">METAL: Fast and efficient meta-analysis of genomewide association scans</a>. <em>Bioinformatics</em> <strong>26</strong>, 2190–1 (2010).</div>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
